sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -67,524 +124,460 @@ static sycl::detail::ESIMDEmuPluginOpaqueData *PiESimdDeviceAccess;
+    std::cerr << " / Line : " << __LINE__ << std::endl;                        \
+  }
 
 pi_result piPlatformsGet(pi_uint32 NumEntries, pi_platform *Platforms,
                          pi_uint32 *NumPlatforms) {
-  (void)NumEntries;
-  (void)Platforms;
-  (void)NumPlatforms;
-  DIE_NO_IMPLEMENTATION;
+
+  static const char *PiTrace = std::getenv("SYCL_PI_TRACE");
+  static const int PiTraceValue = PiTrace ? std::stoi(PiTrace) : 0;
+  if (PiTraceValue == -1) { // Means print all PI traces
+    PrintPiTrace = true;
+  }
+
+  if (NumEntries == 0 && Platforms != nullptr) {
+    return PI_INVALID_VALUE;
+  }
+  if (Platforms == nullptr && NumPlatforms == nullptr) {
+    return PI_INVALID_VALUE;
+  }
+
+  if (Platforms && NumEntries > 0) {
+    *Platforms = new _pi_platform();
+    Platforms[0]->CmEmuVersion = std::string("0.0.1");
+  }
+
+  if (NumPlatforms) {
+    *NumPlatforms = 1;
+  }
+
   return PI_SUCCESS;
 }
 
 pi_result piPlatformGetInfo(pi_platform Platform, pi_platform_info ParamName,
                             size_t ParamValueSize, void *ParamValue,
                             size_t *ParamValueSizeRet) {
-  (void)Platform;
-  (void)ParamName;
-  (void)ParamValueSize;
-  (void)ParamValue;
-  (void)ParamValueSizeRet;
-  DIE_NO_IMPLEMENTATION;
+  if (Platform == nullptr) {
+    return PI_INVALID_PLATFORM;
+  }
+  ReturnHelper ReturnValue(ParamValueSize, ParamValue, ParamValueSizeRet);
+
+  switch (ParamName) {
+  case PI_PLATFORM_INFO_NAME:
+    return ReturnValue("Intel(R) ESIMD_CPU/GPU");
+
+  case PI_PLATFORM_INFO_VENDOR:
+    return ReturnValue("Intel(R) Corporation");
+
+  case PI_PLATFORM_INFO_VERSION:
+    return ReturnValue(Platform->CmEmuVersion);
+
+  case PI_PLATFORM_INFO_PROFILE:
+    return ReturnValue("FULL_PROFILE");
+
+  case PI_PLATFORM_INFO_EXTENSIONS:
+    return ReturnValue("");
+
+  default:
+    // TODO: implement other parameters
+    die("Unsupported ParamName in piPlatformGetInfo");
+  }
+
   return PI_SUCCESS;
 }
 
 pi_result piextPlatformGetNativeHandle(pi_platform, pi_native_handle *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piextPlatformCreateWithNativeHandle(pi_native_handle, pi_platform *) {
   DIE_NO_IMPLEMENTATION;
-  return PI_SUCCESS;
 }
 
 pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
                        pi_uint32 NumEntries, pi_device *Devices,
                        pi_uint32 *NumDevices) {
-  (void)Platform;
-  (void)DeviceType;
-  (void)NumEntries;
-  (void)Devices;
-  (void)NumDevices;
-  DIE_NO_IMPLEMENTATION;
+  if (Platform == nullptr) {
+    return PI_INVALID_PLATFORM;
+  }
+
+  if (NumDevices) {
+    *NumDevices = 1;
+  }
+
+  cm_support::CmDevice *CmDevice = nullptr;
+  unsigned int Version = 0;
+
+  int Result = cm_support::CreateCmDevice(CmDevice, Version);
+
+  if (Result != cm_support::CM_SUCCESS) {
+    return PI_INVALID_DEVICE;
+  }
+
+  if (Devices) {
+    Devices[0] = new _pi_device(Platform, CmDevice);
```

Comment:
If I remember correctly, plugins should return the same device handle for root devices, so some caching is needed here.
Also for root devices `piDeviceRetain` and `piDeviceRelease` should be no-ops. Please, see level zero plugin implementation as a reference. It's OK if this is addressed in a separate PR.

Reply:
> If I remember correctly, plugins should return the same device handle for root devices, so some caching is needed here.
> Also for root devices `piDeviceRetain` and `piDeviceRelease` should be no-ops. Please, see level zero plugin implementation as a reference. It's OK if this is addressed in a separate PR.

It looks like CM does not support 'subdevice'. I'll revise device-related implementations to support only single-root-device. @romanovvlad , let me know if you have different suggestion.